grpc/src/main/java/com/linecorp/armeria/server/grpc/ProtobufRequestConverterFunction.java

```diff
@@ -0,0 +1,211 @@
+                    resultType = ResultType.SET_PROTOBUF;
+                }
+            }
+
+            if (resultType == ResultType.LIST_PROTOBUF || resultType == ResultType.SET_PROTOBUF) {
+                final Class<?> typeArgument =
+                        (Class<?>) expectedParameterizedResultType.getActualTypeArguments()[0];
+                if (Message.class.isAssignableFrom(typeArgument)) {
+                    final String content = request.content(contentType.charset(StandardCharsets.UTF_8));
+
+                    final JsonNode jsonNode = mapper.readTree(content);
+                    if (jsonNode.isArray()) {
+                        final ImmutableCollection.Builder<Message> builder;
+                        if (resultType == ResultType.LIST_PROTOBUF) {
+                            builder = ImmutableList.builderWithExpectedSize(jsonNode.size());
+                        } else {
+                            builder = ImmutableSet.builderWithExpectedSize(jsonNode.size());
+                        }
+                        for (JsonNode node : jsonNode) {
+                            final Message.Builder messageBuilder = getMessageBuilder(typeArgument);
+                            jsonParser.merge(mapper.writeValueAsString(node), messageBuilder);
+                            builder.add(messageBuilder.build());
+                        }
+
+                        return builder.build();
+                    }
+                }
+            }
+        }
+
+        return RequestConverterFunction.fallthrough();
+    }
+
+    private static boolean isJson(@Nullable MediaType contentType) {
+        return contentType != null &&
+               (contentType.is(MediaType.JSON) || contentType.subtype().endsWith("+json"));
+    }
+
+    private static Message.Builder getMessageBuilder(Class<?> clazz) {
+        final MethodHandle methodHandle = methodCache.computeIfAbsent(clazz, key -> {
+            try {
+                final Class<?> builderClass = Class.forName(key.getName() + "$Builder");
+                final Lookup publicLookup = MethodHandles.publicLookup();
+                return publicLookup.findStatic(key, "newBuilder", methodType(builderClass));
+            } catch (NoSuchMethodException | IllegalAccessException | ClassNotFoundException ignored) {
+                return unknownMethodHandle;
+            }
+        });
+        if (methodHandle == unknownMethodHandle) {
+            throw new IllegalStateException("Failed to find a static newBuilder() method from " + clazz);
+        }
+        try {
+            return (Message.Builder) methodHandle.invoke();
+        } catch (Throwable throwable) {
+            throw new IllegalStateException(
+                    "Failed to create an empty instance of " + clazz + " using newBuilder() method", throwable);
+        }
+    }
+
+    enum ResultType {
+        UNKNOWN,
+        PROTOBUF,
+        LIST_PROTOBUF,
+        SET_PROTOBUF
```

Comment:
Do we need to support map as well?

Reply:
`MAP_PROTOBUF` is added for JSON format.